### PR TITLE
Surface GPUI scenario names on panic (10.1.4)

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch implements roadmap task **10.1.4** so failing GPUI scenarios surface
human-readable scenario context without requiring developers to cross-reference
libtest names by hand.

Roadmap task: (10.1.4)
ExecPlan: [docs/execplans/10-1-4-failing-gpui-scenarios-include-scenario-name-in-logs.md](https://github.com/leynos/rstest-bdd/blob/10-1-4-failing-gpui-scenarios-include-scenario-name-in-logs/docs/execplans/10-1-4-failing-gpui-scenarios-include-scenario-name-in-logs.md)

## Review walkthrough

- Start with [crates/rstest-bdd-harness-gpui/src/gpui_harness.rs](https://github.com/leynos/rstest-bdd/blob/10-1-4-failing-gpui-scenarios-include-scenario-name-in-logs/crates/rstest-bdd-harness-gpui/src/gpui_harness.rs) to review the closure-local `catch_unwind` boundary, augmented panic payload, `tracing::error!` record, and stderr diagnostic.
- Then review [crates/rstest-bdd-harness-gpui/tests/scenario_name_in_logs.rs](https://github.com/leynos/rstest-bdd/blob/10-1-4-failing-gpui-scenarios-include-scenario-name-in-logs/crates/rstest-bdd-harness-gpui/tests/scenario_name_in_logs.rs) for the feature-gated regression covering the success path, the augmented failing path, and fresh context after a caught panic.
- Check [crates/rstest-bdd-harness-gpui/Cargo.toml](https://github.com/leynos/rstest-bdd/blob/10-1-4-failing-gpui-scenarios-include-scenario-name-in-logs/crates/rstest-bdd-harness-gpui/Cargo.toml) for the new test target and existing workspace dependencies promoted for production diagnostics.
- Finish with [docs/users-guide.md](https://github.com/leynos/rstest-bdd/blob/10-1-4-failing-gpui-scenarios-include-scenario-name-in-logs/docs/users-guide.md), [docs/developers-guide.md](https://github.com/leynos/rstest-bdd/blob/10-1-4-failing-gpui-scenarios-include-scenario-name-in-logs/docs/developers-guide.md), [docs/rstest-bdd-design.md](https://github.com/leynos/rstest-bdd/blob/10-1-4-failing-gpui-scenarios-include-scenario-name-in-logs/docs/rstest-bdd-design.md), and [docs/roadmap.md](https://github.com/leynos/rstest-bdd/blob/10-1-4-failing-gpui-scenarios-include-scenario-name-in-logs/docs/roadmap.md) for the public documentation and roadmap completion note.

## Validation

- `cargo test -p rstest-bdd-harness-gpui --features native-gpui-tests --test scenario_name_in_logs`: 3 passed.
- `cargo test -p rstest-bdd-harness-gpui --features native-gpui-tests`: 27 passed.
- `make check-fmt`: passed.
- `make lint`: passed.
- `make test`: 1453 nextest tests passed, 7 skipped; 62 Python release automation tests passed.
- `make markdownlint`: passed.
- `coderabbit review --agent`: passed twice with `findings: 0` after the behavioural and documentation milestones.

## Notes

`make fmt` currently fails in its Markdown auto-fix phase on unrelated documents
even though `make markdownlint` passes after reverting that formatter churn. The
ExecPlan records this caveat and the exact transcript locations under `/tmp`.

Lody session: https://lody.ai/leynos/sessions/7edcb31f-104d-44c8-89aa-e2a4baa2f596

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai review

File: docs/users-guide.md [hunk 1]
```diff
@@ -1996,18 +2041,61 @@ https://docs.rs/i18n-embed/latest/i18n_embed/fluent/struct.FluentLanguageLoader.
...
+} = err
+else {
+    panic!("unexpected error variant");
+};
+assert_eq!(row_number, 2);
+assert_eq!(column_index, 2);
+assert!(err
+    .to_string()
+    .contains("unrecognised boolean value 'maybe'"));
 ```
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (4 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 25.00% which is insufficient. The required threshold is 80.00%. | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| Developer Documentation | ⚠️ Warning | Test targets `harness_behaviour` and `attribute_policy_behaviour` are defined in Cargo.toml but missing from the developers-guide test organisation table. | Add `harness_behaviour` and `attribute_policy_behaviour` rows to the test organisation table in docs/developers-guide.md for both Tokio and GPUI harness crates. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Tests exercise the functional boundary but lack edge-case and error-path coverage; stderr failures, special characters, and panic invariants untested. | Add tests for error paths (stderr write failures), edge cases (special characters in names), panic message invariants, and teardown panic handling to improve robustness. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Test uses generic contains() assertions for deterministic panic message, tracing, and stderr output instead of snapshot testing with field redaction. | Add insta snapshot tests for the augmented panic format, using explicit redaction of variable fields like scenario name to keep snapshots meaningful. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

File: docs/developers-guide.md [hunk 2]
```diff
@@ -74,24 +74,29 @@ Mitigation:
...
+| `rstest-bdd-harness-tokio` | `scenario_macros`            | `#[scenario]` + Tokio adapter                                        |
+| `rstest-bdd-harness-tokio` | `macro_compile`              | trybuild compile-pass/fail for Tokio fixtures                        |
+| `rstest-bdd-harness-gpui`  | `harness_behaviour`          | GPUI harness adapter execution semantics (feature-gated)             |
+| `rstest-bdd-harness-gpui`  | `attribute_policy_behaviour` | GPUI attribute policy output (feature-gated)                         |
+| `rstest-bdd-harness-gpui`  | `scenario_macros`            | `#[scenario]` + GPUI adapter (feature-gated)                         |
+| `rstest-bdd-harness-gpui`  | `stateful_window`            | durable GPUI handles + visual context reconstruction (feature-gated) |
+| `rstest-bdd-harness-gpui`  | `scenario_name_in_logs`      | GPUI step-panic diagnostics include scenario context (feature-gated) |
+| `rstest-bdd-harness-gpui`  | `macro_compile`              | trybuild compile-pass for GPUI fixtures (feature-gated)              |
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/developers-guide.md:133]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (4 errors, 6 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Code violates two review requirements: mem::forget missing before resume\_unwind; stderr-write-failure log lacks scenario metadata fields. | Add mem::forget(payload) before panic::resume\_unwind. Add feature\_path, scenario\_name, scenario\_line to stderr-write-failure debug log. |
| Unit Architecture | ❌ Error | Stderr write error handler lacks scenario context fields; payload destructor not forgotten before resume\_unwind risks double-panic abort. | Add scenario context fields to stderr debug log (lines 135–139); add std::mem::forget(payload) before resume\_unwind (line 141) to prevent payload destructor panics. |
| Security And Privacy | ❌ Error | Two unaddressed review comments: (1) stderr-write-failure log omits scenario metadata (feature\_path, scenario\_name, scenario\_line); (2) mem::forget(payload) missing before panic::resume\_unwind. | Add feature\_path, scenario\_name, scenario\_line to line 135-139 debug! call; insert std::mem::forget(payload) before line 141 resume\_unwind. |
| Rust Compiler Lint Integrity | ❌ Error | #\[expect(dead\_code)\] in scenario\_name\_in\_logs.rs lacks required tracking reference; reason must link to PR, issue, or roadmap task per check requirements. | Update CustomPayload #\[expect\] reason to include explicit tracking reference, e.g. "field only exists to produce an opaque Any payload; see PR `#496`". |
| Testing (Unit And Behavioural) | ⚠️ Warning | Two review-comment issues are not addressed: mem::forget is missing before resume\_unwind; tracing::debug! in stderr error path lacks feature\_path, scenario\_name, scenario\_line fields. | Add mem::forget(payload) before resume\_unwind. Add feature\_path, scenario\_name, scenario\_line to the debug log in the stderr write failure path. Add tests verifying both changes. |
| Domain Architecture | ⚠️ Warning | Two review defects prevent domain architecture from being correct: stderr-write failure log lacks scenario context; mem::forget(payload) missing before resume\_unwind. | Add metadata fields to stderr-write failure debug log. Add mem::forget before panic::resume\_unwind to prevent double-panic from original payload destructor. |
| Observability | ⚠️ Warning | Debug log on stderr write failure omits scenario context fields. Catch\_unwind payload not forgotten before resume\_unwind, risking double-panic. | Add feature\_path, scenario\_name, scenario\_line to debug! event. Insert mem::forget(payload) before panic::resume\_unwind. |
| Performance And Resource Use | ⚠️ Warning | Debug log for stderr write failure lacks scenario metadata fields; resume\_unwind lacks mem::forget risking double-panic. | Add feature\_path, scenario\_name, scenario\_line to debug statement at line 135-139. Add mem::forget(payload) before resume\_unwind at line 141. |
| Concurrency And State | ⚠️ Warning | Panic payload not neutralised before resume\_unwind, risking double-panic abort. Debug log for stderr-write failures missing scenario context fields. | Add std::mem::forget(payload) before panic::resume\_unwind; attach feature\_path, scenario\_name, scenario\_line to stderr-write debug event. |
| Architectural Complexity And Maintainability | ⚠️ Warning | Two outstanding review issues: stderr-failure debug log lacks feature\_path/scenario\_name/scenario\_line fields; mem::forget(payload) missing before resume\_unwind risks double-panic. | Add feature\_path, scenario\_name, scenario\_line to tracing::debug! at lines 135–139; add std::mem::forget(payload); before panic::resume\_unwind at line 141. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[crates/rstest-bdd-harness-gpui/tests/scenario_name_in_logs.rs](https://github.com/leynos/rstest-bdd/pull/496/files/ee7501d1d817a318c19813cd86e071197484f486#diff-59550db4675cc4d00bfa202e1ee17c35104b385ee1542fe004d7e38263064528)

Comment on lines +49 to +59

~~~rust
fn failing_scenario_diagnostic_includes_scenario_name() {
    let request = ScenarioRunRequest::new(
        scenario_metadata(FAILING_SCENARIO),
        ScenarioRunner::new(|_context: gpui::TestAppContext| {
            panic!("{STEP_PANIC}");
        }),
    );

    let message = catch_scenario_panic(request);
    configured_snapshot_settings().bind(|| insta::assert_snapshot!(&message));
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69391/delta?repo-id=1163913&review-id=496&biomarker=Code+Duplication&filename=crates%2Frstest-bdd-harness-gpui%2Ftests%2Fscenario_name_in_logs.rs&method=)
The module contains 3 functions with similar structure: augmented_message_includes_scenario_name_for_str_payload,augmented_message_includes_scenario_name_for_string_payload,failing_scenario_diagnostic_includes_scenario_name

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 errors, 1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Unit Architecture | ❌ Error | run\_request\_once violates SRP by bundling runner orchestration, panic handling, message augmentation, tracing, and I/O. Hard-coded io::stderr() lacks dependency injection. | Extract panic observability into injected callback; delegate augmentation/tracing/stderr to caller. Inject stderr sink instead of hard-coding io::stderr(). |
| Rust Compiler Lint Integrity | ❌ Error | The #\[expect(dead\_code)\] on CustomPayload lacks a tracking reference; reason must cite a GitHub issue or tracked item per lint integrity requirements. | Update reason to include a tracking reference: "CustomPayload variant test for opaque Any panic payloads (see ExecPlan 10.1.4`)". |
| Testing (Unit And Behavioural) | ⚠️ Warning | write\_stderr\_diagnostic\_to\_returns\_err\_on\_io\_failure tests private helper via mock, violating the requirement against disguising unit tests as behavioural tests. | Integrate I/O error testing into integration tests through subprocess scenarios or remove and test the edge case differently. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for the following concern and provide an AI coding agent prompt for the fix:

Concurrency And State ⚠️ Warning Tests in scenario\_name\_in\_logs.rs lack #\[serial\] markers required by ExecPlan 10.1.4. GPUI uses Rc-based non-thread-safe entities; concurrent parallel execution of GPUI tests would cause data races. Add #\[serial\] marker from serial\_test crate to each test function in crates/rstest-bdd-harness-gpui/tests/scenario\_name\_in\_logs.rs; import serial\_test::serial.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Unit And Behavioural) | ⚠️ Warning | PR removed the BrokenPipeWriter unit test for error handling during refactoring. Unit tests lack coverage of local behaviour: panic-path helpers not tested in isolation. | Add unit tests in gpui\_harness.rs for augmented\_panic\_message, record\_panic\_event, and write\_stderr\_diagnostic\_to error handling to verify local behaviour and edge cases. |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[crates/rstest-bdd-harness-gpui/src/gpui_harness.rs](https://github.com/leynos/rstest-bdd/pull/496/files/b0b5daf49165151d83314a3b9b14112049b7b214#diff-f80a1e75031ea54096bbc81ea0166f08f3e6cef4b8ac6db352a9ac138b9b8f72)

Comment on lines +353 to +366

~~~rust
    fn augmented_panic_message_includes_scenario_name_for_string_payload() {
        let metadata = ScenarioMetadata::new(
            "tests/features/example.feature",
            "A string payload scenario",
            7,
            vec![],
        );
        let payload: Box<dyn std::any::Any + Send> = Box::new("step panicked".to_string());
        let message = GpuiHarness::augmented_panic_message(payload.as_ref(), &metadata);
        assert!(
            message.contains("A string payload scenario"),
            "expected scenario name in: {message}"
        );
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69391/delta?repo-id=1163913&review-id=496&biomarker=Code+Duplication&filename=crates%2Frstest-bdd-harness-gpui%2Fsrc%2Fgpui_harness.rs&method=)
The module contains 3 functions with similar structure: tests.augmented_panic_message_includes_scenario_name_for_opaque_payload,tests.augmented_panic_message_includes_scenario_name_for_str_payload,tests.augmented_panic_message_includes_scenario_name_for_string_payload

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review
